### DISCUSSION TITLE
Making the project description more enticing

### DISCUSSION THREAD

**Discussion by tulku** [Ekumen-OS/beluga#176] [General]
The project is really nice and has awesome potential! However, as a potential user finding the repo, the current project description in the README file, does not tell me why I should bother with `beluga`, and this might hinder its adoption.

The description states that it is a drop-in replacement for AMCL. But why is it better / different than AMCL? Are there cases where I should still use AMCL? What do you mean by drop-in? Should I expect the same localization performance, or does it have the same interface?

Besides using it instead of AMCL, how do you envision the code to be used? Creating custom localization solutions based on this library? Could you provide examples of that use case? For example, an 'AMCL-like node where you can use Apriltag detections to weight your particles besides the laser'

After that, and a bit out of the topic of this discussion: If I am tempted, how should I make the migration? Are there tools to compare localization performance between AMCL (or cartographer for example) to beluga? Localization is usually a key component of a robot and is hard to evaluate/test for regressions so having some guidelines there can go a long way to help adoption.

@nahueespinosa I believe this suggestion would be mostly of interest to you.

**Comment by nahueespinosa** [comment]
Hello! First off, @tulku thanks a lot for your input!

I believe there are some changes we can make to the current version of docs to address some things and I'll also leave some thoughts here.

FYI @hidmic @glpuga

> But why is it better / different than AMCL?

I tried adding a few extra lines to the top-level readme in #186 about why it is good. I accept suggestions to express it better.

> Are there cases where I should still use AMCL?

We don't know that, and I'm not sure if we want to make any recommendations.

> What do you mean by drop-in? Should I expect the same localization performance, or does it have the same interface?

It should be clear after #186 that we are referring to interface compatibility. It's a bit early to claim anything related to localization performance IMO.

> Besides using it instead of AMCL, how do you envision the code to be used? Creating custom localization solutions based on this library? Could you provide examples of that use case? For example, an 'AMCL-like node where you can use Apriltag detections to weight your particles besides the laser'

I think the API needs to be a bit more mature to start providing examples. #170 and #171 will help a lot in that sense, so I'd wait a little before addressing this question.

> How should I make the migration? Are there tools to compare localization performance between AMCL (or cartographer for example) to beluga? Localization is usually a key component of a robot and is hard to evaluate/test for regressions so having some guidelines there can go a long way to help adoption.

Currently we have a `beluga_benchmark` package that addresses this, after #186 it should be more visible. In the future, #163 will give us excellent tools to deal with this problem.

**Comment by hidmic** [comment]
@nahueespinosa Looks like some debt remains here.

**Reply by nahueespinosa** [reply]
Yeah, I added the topic to the agenda, maybe we can brainstorm about this next week.

**Comment by MoffKalast** [comment]
> I tried adding a few extra lines to the top-level readme in https://github.com/Ekumen-OS/beluga/pull/186 about why it is good. I accept suggestions to express it better.

Spotted this package on the ROS discourse weekly digest, but I'm really not sure what this is all about even with the readme update. Doesn't Nav2 already have an AMCL node? Is this easier to include in rclcpp nodes for customization, and using it without a standalone node to reduce traffic?

**Reply by nahueespinosa** [reply]
This is a somewhat old thread. We're about to start an effort to improve the documentation (as well as add example use cases) to make things clearer for people who find the repository like you did. Admittedly, that effort hasn't started yet due to resource availability, but it should start soon! You'll find tickets related to this in the repository's issue tracker.

To try to answer your questions:

> Doesn't Nav2 already have an AMCL node?

Yes, and it has been stable for many years now, but there are no plans (that we know of) to implement the more recent developments around Monte Carlo methods for Bayesian inference (such as Markov Chain Monte Carlo variants, particle flow filters, hybrid filters using DNNs and RNNs, and more). Our intention is to make Beluga a platform to test these new ideas without people having to start from scratch every time.

As of today, the node has some interesting additional features that Nav2 AMCL does not have:

- Parallelization support that can reduce latency.
- The library itself is ROS-agnostic, so it can potentially be used in a variety of applications.
- It is highly modular and extensible and each component is tested individually.

> Is this easier to include in rclcpp nodes for customization, and using it without a standalone node to reduce traffic?

We're aiming for that. Currently the user-facing API is going through a rewrite that should make it really easy to use different components of the library directly.

**Reply by MoffKalast** [reply]
Ah thanks for the explanation, I think that about covers it. Sounds great!
